MRG: Add whitening to plots

larsoner opened the pull request:
Need to:

- [x] fill out docstrings
- [x] add tests
- [x] update whats_new.rst

Will post CircleCI result once it comes up.

Closes #3926.

larsoner commented:
Certainly not the prettiest example I've ever made:

https://6600-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.AXJHm7i/html/auto_tutorials/plot_whitened.html

larsoner commented:
cc @olafhauk feel free to try this out to see if it addresses your use case from #3926

larsoner commented:
This is now ready for review/merge from my end

agramfort commented:
see texts overlap

https://6600-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.AXJHm7i/html/_images/sphx_glr_plot_whitened_009.png

agramfort commented:
if you pass a noise cov that does not contain a certain channel you should remove it as presently when I click on a bad channel I get errors such as:

```
  File "/Users/alex/work/src/mne-python/mne/cov.py", line 1639, in compute_whitener
    % missing)
RuntimeError: Not all channels present in noise covariance:
['MEG 2443']
```

larsoner commented:
@agramfort fixed, now if you un-mark a bad it disappears from the plot, feel free to try it. I also fixed the overlapping name issue in the example, and a small bug with switching to/from `butterfly` mode.

Here are the updated examples:

https://6665-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.O3jInQw/html/auto_tutorials/plot_whitened.html

https://6665-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.O3jInQw/html/auto_tutorials/plot_visualize_raw.html

agramfort commented:
@larsoner you need to rebase this one

agramfort commented on mne/viz/epochs.py:
```diff
@@ -750,6 +749,12 @@ def plot_epochs(epochs, picks=None, scalings=None, n_epochs=20, n_channels=20,
+    noise_cov : instance of Covariance | str | None
+        Noise covariance used to whiten the data while plotting.
+        Whitened data channels are scaled by ``scalings['whitened']``,
+        and their channel names are shown in italic.
```
why the str option?

agramfort replied:
now I see that it is to be able to read from file from disk.
Please document it.

agramfort commented on mne/viz/tests/test_evoked.py:
```diff
@@ -78,181 +78,184 @@ def test_plot_evoked():
     import matplotlib.pyplot as plt
     rng = np.random.RandomState(0)
     evoked = _get_epochs().average()
```
can you explain me the big diff in this file?

larsoner replied:
There is no reason to have the `with warnings.catch_warnings` in there -- it could be masking useful warnings (and might have been, don't remember)

agramfort commented:
if I toggle off the bad eeg channel in the epochs.plot I get:

```
Traceback (most recent call last):
  File "/Users/alex/anaconda/envs/mne/lib/python3.6/site-packages/matplotlib/cbook/__init__.py", line 389, in process
    proxy(*args, **kwargs)
  File "/Users/alex/anaconda/envs/mne/lib/python3.6/site-packages/matplotlib/cbook/__init__.py", line 227, in __call__
    return mtd(*args, **kwargs)
  File "/Users/alex/work/src/mne-python/mne/viz/epochs.py", line 1490, in _mouse_click
    params['label_click_fun'](pos)
  File "/Users/alex/work/src/mne-python/mne/viz/epochs.py", line 1462, in _pick_bad_channels
    params['plot_update_proj_callback'](params)
  File "/Users/alex/work/src/mne-python/mne/viz/epochs.py", line 1342, in _plot_update_epochs_proj
    params['info'], params['noise_cov'], True, params['use_noise_cov'])
  File "/Users/alex/work/src/mne-python/mne/viz/utils.py", line 2269, in _setup_plot_projector
    noise_cov, use_info, verbose=False)
  File "<string>", line 2, in compute_whitener
  File "/Users/alex/work/src/mne-python/mne/utils.py", line 727, in verbose
    return function(*args, **kwargs)
  File "/Users/alex/work/src/mne-python/mne/cov.py", line 1640, in compute_whitener
    % missing)
RuntimeError: Not all channels present in noise covariance:
['EEG 053']
```

also when I do evoked.plot(noise_cov=noise_cov) and select a time interval the topos that are produced look weird.

larsoner commented on mne/viz/tests/test_evoked.py:
```diff
@@ -78,181 +78,184 @@ def test_plot_evoked():
-        # test GFP only
-        evoked.plot(gfp='only')
-        assert_raises(ValueError, evoked.plot, gfp='foo')
+    evoked.plot(noise_cov=cov_fname)
```
this is the new line

larsoner commented:
> when I do evoked.plot(noise_cov=noise_cov) and select a time interval the topos that are produced look weird.

I think this likely isn't bad math, but rather just whitened topos looking weird.

larsoner commented:
> if I toggle off the bad eeg channel in the epochs.plot I get [an error]

Actually I think this is reasonable behavior and error message. If you try to plot whitened data and your noise cov does not have all data channels, you should get an error, no?

Other comments addressed and rebased.

agramfort commented:
I think we should not display signals that are not part of the noise
covariance. That would be the easiest.

larsoner commented:
Okay lines have been removed

larsoner commented:
CI failure is a sporadic unrelated Travis OSX problem

dengemann commented:
@larsoner @agramfort why would we need evoked.plot(noise_cov=noise_cov) if there is evoked.plot_white? How should a user understand when to use which one? I am not sure what is the answer.

larsoner commented:
The former is (consistent with raw and epochs options and) more useful for
browsing data and responses, the latter is more useful for diagnostic plots
for the covariance.

larsoner commented:
Any chance for review here? I'd like to move on to doing `group_by` support in `epochs.plot` but don't want to start until this is in to avoid conflicts.

agramfort commented:
just played a bit. Behavior with bad channels is still not super robust.

See

[image: Inline image 1]

Basically you see that the bad channel in whiten mode is not plotted but
the label is not in Italic. You can also click on the bad in whiten mode
and you
get a gray line although in whiten mode the bad channels should just be
not plotted at all.

larsoner commented:
@agramfort done. (Also the image you tried to post did not post, but I think I understood what you meant: never show traces of bads from `info` or `cov` in whiten mode.)

agramfort commented:
you should disable toggling bad channels when clicking on channel name in
whitened mode.

marking a good channel as bad in whitened mode gives:

Traceback (most recent call last):
  File
"/Users/alex/anaconda/envs/mne/lib/python3.6/site-packages/matplotlib/cbook/__init__.py",
line 389, in process
    proxy(*args, **kwargs)
  File
"/Users/alex/anaconda/envs/mne/lib/python3.6/site-packages/matplotlib/cbook/__init__.py",
line 227, in __call__
    return mtd(*args, **kwargs)
  File "/Users/alex/work/src/mne-python/mne/viz/epochs.py", line 1491, in
_mouse_click
    params['label_click_fun'](pos)
  File "/Users/alex/work/src/mne-python/mne/viz/epochs.py", line 1463, in
_pick_bad_channels
    params['plot_update_proj_callback'](params)
  File "/Users/alex/work/src/mne-python/mne/viz/epochs.py", line 1343, in
_plot_update_epochs_proj
    params['info'], params['noise_cov'], True, params['use_noise_cov'])
  File "/Users/alex/work/src/mne-python/mne/viz/utils.py", line 2278, in
_setup_plot_projector
    for pick in good_picks]
AssertionError

larsoner commented:
I think it would we unexpected not to be able to toggle bad channels when
in whitening mode. It violates "least surprise" to me at least, and removes
functionality without any real reason to do so.

I'd rather restore the "bad as gray, not in covariance as transparent"
functionality that was there before (and fix the errors of course). Then
users can toggle bads regardless of mode.

agramfort commented:
>
> I think it would we unexpected not to be able to toggle bad channels when
> in whitening mode. It violates "least surprise" to me at least, and removes
> functionality without any real reason to do so.
>
> I'd rather restore the "bad as gray, not in covariance as transparent"
> functionality that was there before (and fix the errors of course). Then
> users can toggle bads regardless of mode.


ok but you have 4 types of channels (bad and in cov, bad and not in cov,
good in cov, good not in cov)

we need to make sure it works for the 4 cases.

larsoner commented:
I will treat "bad and in cov" as "not in cov", since in practice you don't usually reenable these. Then it's simpler, and the interactive experience should mirror what you could get by setting `info['bads']` in code (which is the same way it works when whitening is off).

larsoner commented:
(sorry I mean treat "bad and in cov['bads']" as "not in cov")

agramfort commented:
sounds like a plan

dengemann commented:
I am -0.5 against the whitening arg in plot_evoked. I am afraid it dilutes
API. Users may get confused and miss that there is plot_evoked_white. What
is the motivation here?

On Sun 21 Jan 2018 at 19:55, Alexandre Gramfort <notifications@github.com>
wrote:

> sounds like a plan
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/4799#issuecomment-359271140>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB0firdn-SP36MqBbJvGzID7mMdKhncxks5tM4g3gaJpZM4QxHRs>
> .
>

larsoner commented:
@agramfort done

@dengemann:

1. `plot_white` is more useful for diagnostic plots for the covariance. `plot_evoked` is more useful for browsing data and looking at responses (it is less crowded with extra cov-related "stuff").
2. More importantly, though, is consistency. It is an argument for `raw.plot`, `epochs.plot`, and `evoked.plot_topo`, so *not* having it as an argument for `evoked.plot` would be unexpected for users.

larsoner commented:
@dengemann for you:

https://github.com/mne-tools/mne-python/pull/4799/commits/518ff57cea2a97ac8f9d58ae9061e455cd6f8845

dengemann commented:
Let’s just understand a bit the overall concept here. Can you share some
insights on what this is enabling? General purpose / expert user
functionality for visualizing whitening? Curious to learn what is
motivating you.
On Sun 21 Jan 2018 at 21:35, Eric Larson <notifications@github.com> wrote:

> @dengemann <https://github.com/dengemann> for you:
>
> 518ff57
> <https://github.com/mne-tools/mne-python/commit/518ff57cea2a97ac8f9d58ae9061e455cd6f8845>
>
> —
> You are receiving this because you were mentioned.
>
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/4799#issuecomment-359278630>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB0fikPzdIxPMBKmbqv19i_FBo1VDoLrks5tM5-dgaJpZM4QxHRs>
> .
>

larsoner commented:
Read #3926

agramfort commented:
xlabels on epochs viewer shows byte and not str on python 3 so it does not render fine:

<img width="160" alt="screen shot 2018-02-03 at 11 25 52" src="https://user-images.githubusercontent.com/161052/35766262-1bcd400a-08d5-11e8-89be-a05e53dec459.png">

agramfort commented:
with:

```
evoked.plot(noise_cov=noise_cov)
evoked.plot_white(noise_cov=noise_cov)
```

the scales for evoked data do not match:

<img width="1519" alt="screen shot 2018-02-03 at 11 28 27" src="https://user-images.githubusercontent.com/161052/35766272-68ca7e86-08d5-11e8-951b-1a1710da1b92.png">

with plot_white we did the proper scalings so during baseline the white data have actually a std of approx 1. This is not the case with your plot. How did you scale this? also for Raw and Epochs?

larsoner commented:
> How did you scale this? also for Raw and Epochs?


The same way for each, just by doing a dot product with the whitener.

I could look into doing the scaling the same way, but I don't like that it
departs from the "this is how your data are scaled for localization" idea /
equivalence. So maybe this should be a difference between plot and
plot_white with a proper documentation update. (Also if you whiten with the
additional plot_white scaling and an empty room covariance, the traces
might be huge in scale, which isn't as useful for data browsing.)

larsoner commented:
(Actually it is probably also huge with this implementation, so never mind
that bit.)

On Feb 3, 2018 07:12, "Eric Larson" <larson.eric.d@gmail.com> wrote:

> > How did you scale this? also for Raw and Epochs?
>
>
> The same way for each, just by doing a dot product with the whitener.
>
> I could look into doing the scaling the same way, but I don't like that it
> departs from the "this is how your data are scaled for localization" idea /
> equivalence. So maybe this should be a difference between plot and
> plot_white with a proper documentation update. (Also if you whiten with the
> additional plot_white scaling and an empty room covariance, the traces
> might be huge in scale, which isn't as useful for data browsing.)
>

larsoner commented:
> xlabels on epochs viewer shows byte and not str on python 3 so it does not render fine:

... actually I'm not sure how my changes could have caused this as I just change the styling not the content of the lines. Does this happen on `master` for you? It looks like a `matplotlib` bug to me.

larsoner commented:
I can confirm the `b''` behavior happens on `master` as well, so let's not address it here.

So the only issue that I can see is the scaling. Let me know if you think it needs to be the same, or if it's okay to have it be different.

larsoner commented:
Okay scaling fixed / should now match that of `plot_white`. Along the way I refactored/simplified of some of the rank calculations.

I also fixed a bug with `plot_evoked_white`:
```
import mne
fname = mne.datasets.testing.data_path() + '/SSS/test_move_anon_raw_spatiotemporal_4s_sss.fif'
raw = mne.io.read_raw_fif(fname)
raw.set_eeg_reference(projection=True)
events = mne.make_fixed_length_events(raw, 1)
epochs = mne.Epochs(raw, events)
cov = mne.compute_covariance(epochs, tmax=0.)
evoked = epochs.average()
evoked.plot_white(cov)
```
on `master`:

![master](https://user-images.githubusercontent.com/2365790/35876599-bda580b4-0b40-11e8-875c-e7cb8dfe6b52.png)

On this PR it gives better labeling (and more uniform line widths to match `evoked.plot`:

![pr](https://user-images.githubusercontent.com/2365790/35877009-128a0608-0b42-11e8-8dac-b20310d4c432.png)

The interesting thing is that our magnetometer scaling gets a bit far from the +/- 1.96 limits for SSS'ed data. Adding `evoked.plot(noise_cov=cov)` above we get:

![scaling](https://user-images.githubusercontent.com/2365790/35877011-15d4e328-0b42-11e8-9eff-b44b974125af.png)

I think is to be expected, though. It actually exists in `plot_white`, too, although it's harder to see because grad and mag are overlayed. I added a note in the docstrings that this is expected and that `plot_white` can be used for additional inspection.

larsoner commented:
Examples (some didn't change much): [plot_whitened](https://6914-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.0kA2dvG/html/auto_tutorials/plot_whitened.html), [plot_compute_covariance](https://6914-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.0kA2dvG/html/auto_tutorials/plot_compute_covariance.html), [plot_evoked_whitening](https://6914-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.0kA2dvG/html/auto_examples/visualization/plot_evoked_whitening.html), [plot_covariance_whitening_dspm](https://6914-1301584-gh.circle-artifacts.com/0/tmp/circle-artifacts.0kA2dvG/html/auto_examples/inverse/plot_covariance_whitening_dspm.html).

dengemann commented on examples/visualization/plot_evoked_whitening.py:
```diff
@@ -5,9 +5,9 @@
-the assumption of Gaussian white noise from which we expect values around
-0 with less than 2 standard deviations. Covariance estimation and diagnostic
-plots are based on [1]_.
+the assumption of Gaussian white noise during the baseline period.
```
can we specify +/- 1.96 here as expectation?

dengemann replied:
Maybe it's enough below.

dengemann commented on mne/viz/epochs.py:
```diff
@@ -749,7 +748,19 @@ def plot_epochs(epochs, picks=None, scalings=None, n_epochs=20, n_channels=20,
+        and their channel names are shown in italic.
+        Can be a string to load a covariance from disk.
+        See also :meth:`mne.Evoked.plot_white` for additional inspection
+        of noise covariance properties when whitening evoked data.
```
, specifically, for inspection of expected values and model violations ... {or so}

larsoner replied:
I'm wary of adding more text to an already long description here. I'll add it to the text for `plot_evoked_white`

dengemann commented on mne/viz/evoked.py:
```diff
@@ -579,6 +581,17 @@ def plot_evoked(evoked, picks=None, exclude='bads', unit=True, show=True,
+        Can be a string to load a covariance from disk.
+        See also :meth:`mne.Evoked.plot_white` for additional inspection
+        of noise covariance properties when whitening evoked data.
+        For data processed with SSS, the effective dependence between
```
whatever is changed above should also be changed here (model violations / expected values etc).

dengemann commented on mne/viz/evoked.py:
```diff
@@ -839,26 +863,20 @@ def _plot_update_evoked(params, bools):
-    whitened GFPs based on each estimator separately for each sensor type.
-    Instead of numbers of channels the GFP display shows the estimated rank.
-    Note. The rank estimation will be printed by the logger for each noise
-    covariance estimator that is passed.
```
why is this removed? There is a lot of specific information. I think this is not a good idea.

dengemann replied:
Ok. Below. Nevermind.

dengemann commented on mne/viz/evoked.py:
```diff
@@ -931,102 +944,31 @@ def _plot_evoked_white(evoked, noise_cov, scalings=None, rank=None, show=True):
-        rank_list.append(this_rank)
+    evoked.pick_types(meg=True, eeg=True, ref_meg=False, exclude='bads')
+    n_ch_used, rank_list, picks_list, has_sss = _triage_rank_sss(
+        evoked.info, noise_cov, rank, scalings)
```
@larsoner is this call replacing all the work of the helper functions and other lines you removed?

larsoner replied:
Yes it was copy-paste to refactor + bug fixes

dengemann commented on mne/viz/ica.py:
```diff
@@ -775,7 +775,8 @@ def _plot_sources_raw(ica, raw, picks, exclude, start, stop, show, title,
                   n_times=raw.n_times, bad_color=bad_color, picks=picks,
-                  first_time=first_time, data_picks=[], decim=1)
+                  first_time=first_time, data_picks=[], decim=1,
+                  noise_cov=None, whitened_ch_names=())
```
@larsoner why are we doing this actually? Just because we can or because people will need this. How does our whitening interact with ICA?

larsoner replied:
Need to pass the params / API uniformity internally. They shouldn't use it with ICA which is why it's not exposed in the public facing APIs for ICA plotting.

dengemann commented on mne/viz/utils.py:
```diff
@@ -2242,3 +2250,192 @@ def _grad_pair_pick_and_name(info, picks):
+        # (but in gray), but here we don't need to because "projector"
+        # starts out as identity. So all that is left to do is take any
+        # *good* data channels that are not in the noise cov to be NaN
+        nan_names = data_names - (bad_names | cov_names)
```
maybe the case for using set methods for better readability.

larsoner replied:
I think set operations are clearly described using their arithmetic equivalents ...?

dengemann commented:
@larsoner I am not sure it is the best example. Can we plot the whitening of the evoked to see what is going on? As we discussed, the magnetometers are expected to be much lower in amplitude after SSS. Together with gradiometers they make up n_sss_components dimensions, as is shown in the GFP plot of the whitening.
Also to be on the same page, can you describe what was that bug that you fixed? / point to the buggy code you replaced with correct code?

We're getting somewhere, don't give up :) I think we're moving in the right direction .

larsoner commented:
> @larsoner I am not sure it is the best example. Can we plot the whitening of the evoked to see what is going on?

Which example? The script I posted above? The three images in my previous post all come from the same script so it has `plot_white` as well as `evoked.plot(noise_cov=cov)` outputs. The traces are the same you just can't see the mags in the `plot_white` case because they are small relative to the grads. (I checked this by manually `.set_visible(False)`ing the gradiometer lines in the `plot_white` case.)

larsoner commented on mne/viz/utils.py:
```diff
@@ -2242,3 +2250,192 @@ def _grad_pair_pick_and_name(info, picks):
+        # we risk the rank estimates being incorrect (i.e., if the projectors
+        # do not match).
+        info_proj = info.copy()
+        info_proj['projs'] += cov['projs']
```
This is one bugfix -- the estimated rank needs to use the union of projectors to mirror what happens in computing the whitener

larsoner commented on mne/viz/utils.py:
```diff
@@ -2242,3 +2250,192 @@ def _grad_pair_pick_and_name(info, picks):
+                projector = setup_proj(this_info, add_eeg_ref=False)[0]
+                this_C = C[idx]
+                if projector is not None:
+                    this_C = np.dot(np.dot(projector, this_C), projector.T)
```
... and correspondingly this joint projector actually needs to be applied to the covariance matrix to ensure rank estimation works properly

larsoner commented on mne/viz/evoked.py:
```diff
@@ -1081,9 +1023,12 @@ def whitened_gfp(x, rank=None):
-                ax.axhline(hline, color='red', linestyle='--')
+                ax.axhline(hline, color='red', linestyle='--', lw=2)
+            ax.set(title='{0} ({1} channels)'.format(titles_[ch_type],
+                                                     len(picks)))
```
this is a bug / consistency fix (the one I show with a script in a previous comment)

larsoner commented:
@dengemann comments addressed (I think)

larsoner commented:
ping @dengemann let me know if this is good to go for you

dengemann commented:
Looks good. We should move on. I let @agramfort take another look.

larsoner commented:
@agramfort okay for you too?
